scaladoc/src/dotty/tools/scaladoc/site/customTags/LanguagePicker.scala

```diff
@@ -0,0 +1,148 @@
+        "en" :: languagesArg.filterNot(_ == "en")
+      } else {
+        // Otherwise, prepend "en" to the list
+        "en" :: languagesArg
+      }
+    }
+
+    // Early return if the languages argument is empty
+    if (languagesWithEnFirst.isEmpty) {
+      println("Languages argument is empty, not rendering anything.")
+      return ""
+    }
+
+    // Extract language codes from the list
+    val extractedLanguageCodes =
+      languagesWithEnFirst.flatMap(extractLanguageCodes)
+
+    // Load the config using ConfigLoader
+    val config = LanguagePickerTag.getConfigValue
+
+    // Access languages from the config
+    val languagesOpt =
+      config.get[java.util.List[java.util.Map[String, String]]]("languages")
+    val languagesFromConfig = languagesOpt match {
+      case Some(list) => list.asScala.toList.map(_.asScala.toMap)
+      case None =>
+        throw new IllegalArgumentException(
+          "No languages found in configuration"
+        )
+    }
+
+    // Extract the list of codes from the config
+    val configLanguageCodes = languagesFromConfig.map(_("code"))
+
+    // Validate that all languages in extractedLanguageCodes exist in the config
+    val missingLanguages =
+      extractedLanguageCodes.filterNot(configLanguageCodes.contains)
+    if (missingLanguages.nonEmpty) {
+      throw new IllegalArgumentException(
+        s"The following languages are not defined in the configuration: ${missingLanguages.mkString(", ")}"
+      )
+    }
+
+    // Filter the languages from the config based on the extractedLanguageCodes
+    val languages = languagesFromConfig.filter(language =>
+      extractedLanguageCodes.contains(language("code"))
+    )
+
+    // Create the dropdown HTML
+    val dropdown = new StringBuilder
+    dropdown.append(
+      "<select id='language-selector' name='language' onChange='handleLanguageChange(this)' style='font-size: 14px; padding: 5px; border: 1px solid #ccc; border-radius: 3px;'>"
+    )
+    for (language <- languages) {
+      val code = language("code")
+      val name = language("name")
+      dropdown.append(s"<option value='$code'>$name</option>")
+    }
+    dropdown.append(
+      "</select> <script>var availableLanguages = " + languagesWithEnFirst
+        .map("'" + _ + "'")
+        .mkString(
+          "[",
+          ", ",
+          "]"
+        ) + "; console.log('Available languages: ', availableLanguages);</script>"
+    )
+
+    dropdown.toString()
+  }
+
+  private def extractLanguageCodes(languagesArg: String): List[String] = {
+    // Regular expression to match both two-letter codes and hyphenated parts
+    val pattern = """[a-z]{2}(-[a-z]{2})?""".r
+
+    // Find all matches in the string
+    val matches = pattern.findAllIn(languagesArg).toList
+
+    // Ensure that each part is valid and non-empty
+    matches.map(_.trim).filter(_.nonEmpty)
+  }
+}
+
+object LanguagePickerTag {
+  @volatile private var config: Config = null
```

Comment:
question: can we get rid of this shared global state too?

Reply:
you should be able to add a constructor argument to `LanguagePickerTag`